neuralop/layers/coda_blocks.py

```diff
@@ -0,0 +1,417 @@
+            self.mixer_out_normalizer = norm_module(
+                self.mixer_token_codimension)
+
+        else:
+            self.mixer = FNOBlocks(
+                in_channels=codimension_size,
+                out_channels=codimension_size,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = norm_module(codimension_size)
+            self.norm2 = norm_module(codimension_size)
+            self.mixer_out_normalizer = norm_module(codimension_size)
+
+    def compute_attention(self, tokens, batch_size):
+        """
+        Compute the key-query-value variant of the attention matrix for input token functions.
+
+        Parameters
+        ----------
+        tokens : torch.Tensor
+            Input tokens with shape (b * t, d, h, w, ...), where:
+            b is the batch size,
+            t is the number of tokens,
+            d is the token codimension,
+            and h, w, ... are the domain dimensions.
+            Assumes input tokens have been normalized.
+        
+        batch_size : int
+            The size of the batch.
+        """
+
+        k = self.Key(tokens)
+        q = self.Query(tokens)
+        v = self.Value(tokens)
+        assert k.size(
+            1) % self.n_heads == 0, "Number of channels in k, q, and v should be divisible by number of heads"
+
+        # reshape from (b*t) (n*d) h w -> b n t (d*h*w ...)
+        t = k.size(0) // batch_size  # Compute the number of tokens `t`
+        # Computer per head token codimension `d`
+        d = k.size(1) // self.n_heads
+
+        # reshape from (b*t) (n*d) h w ... to b n t d h w ...
+        k = k.view(batch_size, t, self.n_heads, d, *k.shape[-self.n_dim:])
+        q = q.view(batch_size, t, self.n_heads, d, *q.shape[-self.n_dim:])
+        v = v.view(batch_size, t, self.n_heads, d, *v.shape[-self.n_dim:])
+
+        k = torch.transpose(k, 1, 2)
+        q = torch.transpose(q, 1, 2)
+        v = torch.transpose(v, 1, 2)
+        # resahpe
+        k = k.view(batch_size, self.n_heads, t, -1)
+        q = q.view(batch_size, self.n_heads, t, -1)
+        v = v.view(batch_size, self.n_heads, t, -1)
+
+        # attention mechanism
+        dprod = (torch.matmul(q, k.transpose(-1, -2)) /
+                 (np.sqrt(k.shape[-1]) * self.temperature))
+        dprod = F.softmax(dprod, dim=-1)
+
+        attention = torch.matmul(dprod, v)
+
+        # Reshape from (b, n, t, d * h * w) to (b, n, t, d, h, w, ...)
+        attention = attention.view(
+            attention.size(0),
+            attention.size(1),
+            attention.size(2),
+            d,
+            *tokens.shape[-self.n_dim:])
+        attention = torch.transpose(attention, 1, 2)
+        attention = attention.reshape(attention.size(0) * attention.size(1),
+                                   attention.size(2) * attention.size(3),
+                                   *attention.shape[-self.n_dim:])
```

Comment:
I think I'd commented on this before: I find it strange to do a view followed by a transpose and a reshape: why not just one transpose followed by one reshape here?